### ISSUE TITLE
changing config.axial_pos_shape for 'ReformerModelWithLMHead' when fine-tuning 

### ISSUE THREAD

**Issue by D-i-l-r-u-k-s-h-i** [huggingface/transformers#4565]
# ❓ Questions & Help

I'm trying to fine-tune the Reformer for the language generation task and I padded the sequence lengths to be a multiple of least common multiple chunk_length 64, and now I'm asked to pad the sequence to 524288(512 * 1024), which will give me an out of memory error. Thank you.

I would like to know a workaround for this, since the error message also gives an alternative to 'pad_to_max_length', which is 'changing config.axial_pos_shape' and specially since this is known to be a memory efficient transformer.
**A link to original question on Stack Overflow**: [https://stackoverflow.com/questions/61986452/fine-tuning-reformer-gives-out-of-memory-error-when-sequence-length-is-padded-t](url)

**Comment by patrickvonplaten** [comment]
I would not recommend to set `axial_pos_shape` to (512 * 1024). In the notebook I just used that to demonstrate how far the limits can be pushed for Reformer. Half a million token is extremely long and usually unnecessary. 

Make sure you have read and understood how AxialPostionEmbeddings work: https://huggingface.co/transformers/model_doc/reformer.html#axial-positional-encodings . 

For "normal" language modeling it might make much more sense to start from the Reformer-wiken8 model and finetune it: https://huggingface.co/google/reformer-enwik8

**Comment by epetros** [comment]
Greetings,
Would fine tuning https://huggingface.co/google/reformer-enwik8 work normally with run_language_modeling.py script?
Thanks

**Comment by patrickvonplaten** [comment]
Hmm, for the most part but you will have to define your own tokenzer function as can be seen here: https://huggingface.co/google/reformer-enwik8#reformer-language-model-on-character-level-and-trained-on-enwik8

**Comment by patrickvonplaten** [comment]
So instead of sticking the script, I would recommend slightly changing this notebook: https://github.com/patrickvonplaten/notebooks/blob/master/PyTorch_Reformer.ipynb. Instead of creating the dataset by using a tokenizer, you should use the function linked above.

**Comment by epetros** [comment]
Ok, thanks. So the function _flatten_and_tokenize_ (in the notebook) shall be replaced by the _encode_
function (in the enwik8 model card), am I following right?

**Comment by D-i-l-r-u-k-s-h-i** [comment]
> I would not recommend to set `axial_pos_shape` to (512 * 1024). In the notebook I just used that to demonstrate how far the limits can be pushed for Reformer. Half a million token is extremely long and usually unnecessary.
> 
I've been using 'google/reformer-crime-and-punishment' model from [https://huggingface.co/transformers/model_doc/reformer.html#reformermodelwithlmhead](url)

I get this error after I padded the sequence lengths to be a multiple of least common multiple chunk_length 64.

>If training, make sure that config.axial_pos_shape factors: (512, 1024) multiply to sequence length. Got prod((512, 1024)) != sequence_length: 384. You might want to consider padding your sequence length to 524288 or changing config.axial_pos_shape.

So I guess that is because its default set to  (512, 1024), and if so, how can I change it? 
Thank you.

**Comment by patrickvonplaten** [comment]
yeah the google/crime-and-punishment is not a good model for fine-tuning. It assumes you use a sequence length of > 500K tokens, which is not really reasonable.

**Comment by patrickvonplaten** [comment]
> Ok, thanks. So the function _flatten_and_tokenize_ (in the notebook) shall be replaced by the _encode_
> function (in the enwik8 model card), am I following right?

exactly. You should be able to just enwik8 function I linked above. The enwik8 model has a maximum length of ~65K tokens, which is very long but very feasible for reformer.

**Comment by D-i-l-r-u-k-s-h-i** [comment]
> yeah the google/crime-and-punishment is not a good model for fine-tuning. It assumes you use a sequence length of > 500K tokens, which is not really reasonable.

Oh okay. Thank you very much for the clarification. Will try finetuning reformer-enwik8.

**Comment by patrickvonplaten** [comment]
It would be awesome if you could upload your training script here - people seem very interested in it :-)

**Comment by D-i-l-r-u-k-s-h-i** [comment]
@patrickvonplaten, Sure, will do when everything is sorted.

**Comment by epetros** [comment]
> 
> 
> > Ok, thanks. So the function _flatten_and_tokenize_ (in the notebook) shall be replaced by the _encode_
> > function (in the enwik8 model card), am I following right?
> 
> exactly. You should be able to just enwik8 function I linked above. The enwik8 model has a maximum length of ~65K tokens, which is very long but very feasible for reformer.

From the notebook am struggling to adapt the DataCollator, how to define it properly in this context? 
Thanks

**Comment by lucashueda** [comment]
Someone effectivelly fine tune on enwiki8 pre trained model? using colab with P100 gpu i was not able to load model yet due to memory limitation

**Comment by epetros** [comment]
Unfortunately facing the same issue now.

**Comment by patrickvonplaten** [comment]
Can you add a link to your notebook here @lucashueda ?

**Comment by mihaidobri** [comment]
@lucashueda Did you manage to fine-tune enwiki8 pre trained model ? or other datasets ? Would you mind sharing your Colab?

**Comment by mihaidobri** [comment]
@epetros did you manage to perform the fine-tuning ?

**Comment by VikasRajashekar** [comment]
@patrickvonplaten any update on this? 

Os is there a notebook where we can pre-train this model from wiki or huge corpus ourselves and then fine-tune it to downstream tasks?

**Comment by param-jot** [comment]
Hi, I'm trying to fine-tune the ReformerModelWithLMHead (google/reformer-enwik8) for NER. I used the padding sequence length same as in the encode method (max_length = max([len(string) for string in list_of_strings])) along with attention_masks. And I got this error: 

**ValueError:** If training, make sure that config.axial_pos_shape factors: (128, 512) multiply to sequence length. Got prod((128, 512)) != sequence_length: 2248. You might want to consider padding your sequence length to 65536 or changing config.axial_pos_shape.

1) When I changed the sequence length to 65536, my colab session crashed by getting all the inputs of 65536 lengths.  
2) According to the second option(changing config.axial_pos_shape), I cannot change it. 

I would like to know, Is there any chance to change config.axial_pos_shape while fine-tuning the model? Or I'm missing something in encoding the input strings for reformer-enwik8?

Thanks!
